File: src/scheduler.py
import logging
import re
from datetime import datetime, timedelta

from aiogram import Bot
from apscheduler.schedulers.asyncio import AsyncIOScheduler
from apscheduler.triggers.cron import CronTrigger
from apscheduler.triggers.interval import IntervalTrigger

from src.claude_client import ClaudeClient
from src.config import Config
from src.session_store import SessionStore
from src.telegram_bot import _split_message

logger = logging.getLogger(__name__)

MORNING_BRIEFING_PROMPT = (
    "Good morning! Please give me my daily training briefing:\n"
    "1. Check my training plan for today on Intervals.icu — what workout is scheduled?\n"
    "2. Check my recent wellness data (sleep, soreness, HRV, mood).\n"
    "3. Look at my recent training load (ATL/CTL/TSB) and fatigue.\n"
    "4. Based on all this, tell me what I should do today.\n"
    "5. Ask me how I'm feeling and if anything needs adjusting.\n"
    "Keep it concise — this is my morning check-in."
)

ACTIVITY_ANALYSIS_PROMPT = """\
Check Intervals.icu for any activities from the last 24 hours \
(use oldest={since_date}).

{skip_clause}

For EACH new activity, start your response with ANALYZED:<activity_id> on its \
own line (e.g. ANALYZED:i129194330), then provide a detailed analysis:

1. **Activity summary**: type, duration, distance, average HR, average power/pace
2. **Training classification**: what kind of training was this? (recovery, endurance/zone 2, tempo/sweetspot, threshold, VO2max, anaerobic, sprint, strength, etc.). Explain WHY you classified it this way based on the intensity distribution, HR zones, and power/pace data.
3. **Performance assessment**: how well did it go? Compare to recent similar activities. Look at pacing consistency, cardiac drift, power/pace decoupling, RPE vs actual intensity.
4. **Scientific context**: use WebSearch to find relevant exercise science research (peer-reviewed papers, systematic reviews) that supports your analysis. For example, if it was a zone 2 ride, cite research on mitochondrial adaptations; if it was intervals, cite research on the specific protocol's effectiveness.
5. **Key takeaways**: 2-3 actionable insights for future training.

For every scientific claim, include a reference with author, year, journal, and the finding. Format references as a numbered list at the end.

If there are NO new activities to analyze, respond with exactly: NO_NEW_ACTIVITIES
"""


def setup_scheduler(
    config: Config,
    bot: Bot,
    claude: ClaudeClient,
    session_store: SessionStore,
) -> AsyncIOScheduler:
    scheduler = AsyncIOScheduler(timezone=config.timezone)

    async def morning_briefing() -> None:
        for user_id in config.allowed_user_ids:
            logger.info("Sending morning briefing to user %d", user_id)

            session_id = await session_store.get_session(user_id)

            response_text, new_session_id = await claude.send_message(
                MORNING_BRIEFING_PROMPT, session_id
            )

            if new_session_id:
                await session_store.save_session(user_id, new_session_id)

            for chunk in _split_message(response_text):
                await bot.send_message(chat_id=user_id, text=chunk)

            logger.info("Morning briefing sent to user %d", user_id)

    async def activity_check() -> None:
        since_date = (
            datetime.now() - timedelta(hours=24)
        ).strftime("%Y-%m-%dT%H:%M")
        analyzed = await session_store.get_analyzed_activities()
        logger.info(
            "Checking for new activities since %s (skip %d already analyzed)",
            since_date,
            len(analyzed),
        )

        if analyzed:
            skip_clause = (
                "Skip these activity IDs (already analyzed): "
                + ", ".join(sorted(analyzed))
            )
        else:
            skip_clause = ""

        for user_id in config.allowed_user_ids:
            session_id = await session_store.get_session(user_id)

            prompt = ACTIVITY_ANALYSIS_PROMPT.format(
                since_date=since_date,
                skip_clause=skip_clause,
            )
            response_text, new_session_id = await claude.send_message(
                prompt, session_id
            )

            if new_session_id:
                await session_store.save_session(user_id, new_session_id)

            if "NO_NEW_ACTIVITIES" in response_text:
                logger.info("No new activities for user %d", user_id)
                continue

            # Extract analyzed activity IDs from response
            for match in re.finditer(r"ANALYZED:(i\d+)", response_text):
                aid = match.group(1)
                await session_store.mark_activity_analyzed(aid)
                logger.info("Marked activity %s as analyzed", aid)

            # Strip ANALYZED: markers before sending to user
            clean_response = re.sub(
                r"^ANALYZED:i\d+\n?", "", response_text, flags=re.MULTILINE
            ).strip()

            if clean_response:
                logger.info(
                    "Sending activity analysis to user %d", user_id
                )
                for chunk in _split_message(clean_response):
                    await bot.send_message(chat_id=user_id, text=chunk)

    scheduler.add_job(
        morning_briefing,
        CronTrigger(
            hour=config.briefing_hour,
            minute=config.briefing_minute,
        ),
        id="morning_briefing",
        replace_existing=True,
    )

    scheduler.add_job(
        activity_check,
        IntervalTrigger(minutes=30),
        id="activity_check",
        replace_existing=True,
    )

    logger.info(
        "Morning briefing scheduled at %02d:%02d %s",
        config.briefing_hour,
        config.briefing_minute,
        config.timezone,
    )
    logger.info("Activity check scheduled every 30 minutes")

    return scheduler

File: internal/scheduler/scheduler.go
package scheduler

import (
	"context"
	"fmt"
	"log"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/go-co-op/gocron/v2"

	"github.com/derrix060/ai-fitness-trainer/internal/claude"
	"github.com/derrix060/ai-fitness-trainer/internal/config"
	"github.com/derrix060/ai-fitness-trainer/internal/store"
)

var (
	analyzedRe     = regexp.MustCompile(`ANALYZED:(i\d+)`)
	analyzedLineRe = regexp.MustCompile(`(?m)^ANALYZED:i\d+\n?`)
)

// Sender can send formatted messages to a Telegram chat.
type Sender interface {
	SendFormatted(ctx context.Context, chatID int64, text string)
}

func Setup(
	cfg *config.Config,
	sender Sender,
	c *claude.Client,
	s *store.Store,
) (gocron.Scheduler, error) {
	sched, err := gocron.NewScheduler(gocron.WithLocation(cfg.Timezone))
	if err != nil {
		return nil, err
	}

	_, err = sched.NewJob(
		gocron.CronJob(
			fmt.Sprintf("%d %d * * *", cfg.BriefingMinute, cfg.BriefingHour),
			false,
		),
		gocron.NewTask(func() {
			morningBriefing(cfg, sender, c, s)
		}),
	)
	if err != nil {
		return nil, fmt.Errorf("schedule morning briefing: %w", err)
	}

	_, err = sched.NewJob(
		gocron.DurationJob(30*time.Minute),
		gocron.NewTask(func() {
			activityCheck(cfg, sender, c, s)
		}),
	)
	if err != nil {
		return nil, fmt.Errorf("schedule activity check: %w", err)
	}

	log.Printf("Morning briefing scheduled at %02d:%02d %s",
		cfg.BriefingHour, cfg.BriefingMinute, cfg.Timezone.String())
	log.Printf("Activity check scheduled every 30 minutes")

	return sched, nil
}

func morningBriefing(cfg *config.Config, sender Sender, c *claude.Client, s *store.Store) {
	ctx := context.Background()
	for userID := range cfg.AllowedUserIDs {
		log.Printf("Sending morning briefing to user %d", userID)

		sessionID, _ := s.GetSession(userID)
		responseText, newSessionID, err := c.SendMessage(ctx, MorningBriefingPrompt, sessionID)
		if err != nil {
			log.Printf("Morning briefing error for user %d: %v", userID, err)
			continue
		}

		if newSessionID != "" {
			s.SaveSession(userID, newSessionID)
		}
		sender.SendFormatted(ctx, userID, responseText)
		log.Printf("Morning briefing sent to user %d", userID)
	}
}

func activityCheck(cfg *config.Config, sender Sender, c *claude.Client, s *store.Store) {
	ctx := context.Background()
	sinceDate := time.Now().Add(-24 * time.Hour).Format("2006-01-02T15:04")
	analyzed, _ := s.GetAnalyzedActivities()
	log.Printf("Checking for new activities since %s (skip %d already analyzed)",
		sinceDate, len(analyzed))

	var skipClause string
	if len(analyzed) > 0 {
		ids := make([]string, 0, len(analyzed))
		for id := range analyzed {
			ids = append(ids, id)
		}
		sort.Strings(ids)
		skipClause = "\nSkip these activity IDs (already analyzed): " + strings.Join(ids, ", ")
	}

	for userID := range cfg.AllowedUserIDs {
		sessionID, _ := s.GetSession(userID)
		prompt := fmt.Sprintf(ActivityAnalysisPrompt, sinceDate, skipClause)

		responseText, newSessionID, err := c.SendMessage(ctx, prompt, sessionID)
		if err != nil {
			log.Printf("Activity check error for user %d: %v", userID, err)
			continue
		}

		if newSessionID != "" {
			s.SaveSession(userID, newSessionID)
		}

		if strings.Contains(responseText, "NO_NEW_ACTIVITIES") {
			log.Printf("No new activities for user %d", userID)
			continue
		}

		for _, match := range analyzedRe.FindAllStringSubmatch(responseText, -1) {
			s.MarkActivityAnalyzed(match[1])
			log.Printf("Marked activity %s as analyzed", match[1])
		}

		clean := strings.TrimSpace(analyzedLineRe.ReplaceAllString(responseText, ""))
		if clean != "" {
			log.Printf("Sending activity analysis to user %d", userID)
			sender.SendFormatted(ctx, userID, clean)
		}
	}
}
